src/tools/MonarchPeasantSample/main.cpp

```diff
@@ -0,0 +1,175 @@
+﻿#include "pch.h"
+#include <conio.h>
+#include "Monarch.h"
+#include "Peasant.h"
+#include "AppState.h"
+#include "../../types/inc/utils.hpp"
+
+using namespace winrt;
+using namespace winrt::Windows::Foundation;
+using namespace ::Microsoft::Console;
+
+////////////////////////////////////////////////////////////////////////////////
+// This seems like a hack, but it works.
+//
+// This class factory works so that there's only ever one instance of a Monarch
+// per-process. Once the first monarch is created, we'll stash it in g_weak.
+// Future callers who try to instantiate a Monarch will get the one that's
+// already been made.
+//
+// I'm sure there's a better way to do this with WRL, but I'm not familiar
+// enough with WRL to know for sure.
+
+winrt::weak_ref<MonarchPeasantSample::implementation::Monarch> g_weak{ nullptr };
+
+struct MonarchFactory : implements<MonarchFactory, IClassFactory>
+{
+    MonarchFactory() :
+        _guid{ Monarch_clsid } {};
+
+    HRESULT __stdcall CreateInstance(IUnknown* outer, GUID const& iid, void** result) noexcept final
+    {
+        *result = nullptr;
+        if (outer)
+        {
+            return CLASS_E_NOAGGREGATION;
+        }
+
+        if (!g_weak)
+        {
+            // Create a new Monarch instance
+            auto strong = make_self<MonarchPeasantSample::implementation::Monarch>();
+
+            g_weak = (*strong).get_weak();
+            return strong.as(iid, result);
+        }
+        else
+        {
+            // We already instantiated one Monarch, let's just return that one!
+            auto strong = g_weak.get();
+            return strong.as(iid, result);
+        }
+    }
+
+    HRESULT __stdcall LockServer(BOOL) noexcept final
+    {
+        return S_OK;
+    }
+
+private:
+    winrt::guid _guid;
+};
+////////////////////////////////////////////////////////////////////////////////
+
+// Function Description:
+// - Register the Monarch object with COM. This allows other processes to create
+//   Monarch's in our process space with CoCreateInstance and the Monarch_clsid.
+DWORD registerAsMonarch()
+{
+    DWORD registrationHostClass{};
+    check_hresult(CoRegisterClassObject(Monarch_clsid,
```

Comment:
Feels a little dirty to be both Ro registering this because it's a winrt type and Co registering this again with a CLSID GUID... but as far as I can see online... that's the way it works..

Reply:
Yea I mean, it's definitely a little weird. We're basically saying "There's this COM thing, a `Monarch`, that you can get with this GUID". Then the client is saying "I would like one COM-y `Monarch`", to which the server happily obliges. Then the client IMMEDIATELY turns around and uses that thing as a WinRT-y `Monarch`.

I'm thinking of abusing this for the tearout/handles thing, where the object implements both a WinRT interface AND a COM interface (with handles). So then the client can go "I would like a `Foo`" and then decide to do either COM things (handles) or WinRT things (events) depending on the context.

Reply:
NO-GOD-NO.gif

Tried throwing this in a package this morning, as a curiousity. 

Immediately failed with `Error: No such interface supported`. So now I have to figure out what combination of magic manifest incantations I need to add to make it possible to activate these classes when running with package identity. Is it the SxS manifest? Is it in the `Package.appxManifest`? _TUNE IN NEXT TIME_

Reply:
That's definitely what's messing with me here. If I run the sample _unpackaged_, then create another terminal and try to run the _packaged_ exe, I get:
```
Error: The process has no package identity.
```

So somehow the packaged exe knows that the server needs to have package identity? the package identity is definitely the thing that's blocking it though.

Reply:
Is this the bit you solved with the `winmd` file per the email thread this morning?

Reply:
Yes, yes it is.